velox/functions/prestosql/window/Ntile.cpp

```diff
@@ -0,0 +1,225 @@
+    // (number of rows in partition) % (number of buckets) such buckets.
+    int64_t bucketsWithExtraRow;
+    // When assigning bucket numbers, the first 'bucketsWithExtraRow' buckets
+    // will have (rowsPerBucket + 1) rows. This row number at this boundary is
+    // extraBucketsBoundary = bucketsWithExtraRow * (rowsPerBucket + 1). Beyond
+    // this row number in the partition, the buckets will have only
+    // rowsPerBucket number of rows. This boundary is useful when computing the
+    // bucket value.
+    int64_t extraBucketsBoundary;
+  };
+
+  inline const BucketMetrics computeBucketMetrics(int64_t numBuckets) {
+    auto rowsPerBucket = numPartitionRows_ / numBuckets;
+    auto bucketsWithExtraRow = numPartitionRows_ % numBuckets;
+    auto extraBucketsBoundary = (rowsPerBucket + 1) * bucketsWithExtraRow;
+    return {rowsPerBucket, bucketsWithExtraRow, extraBucketsBoundary};
+  }
+
+  inline int64_t computeBucketValue(
+      vector_size_t rowNumber,
+      const BucketMetrics& bucketMetrics) {
+    if (rowNumber < bucketMetrics.extraBucketsBoundary) {
+      return rowNumber / (bucketMetrics.rowsPerBucket + 1) + 1;
+    }
+    return (rowNumber - bucketMetrics.bucketsWithExtraRow) /
+        bucketMetrics.rowsPerBucket +
+        1;
+  }
+
+  void computeNtileFromColumn(
+      vector_size_t numRows,
+      vector_size_t resultOffset,
+      const VectorPtr& result) {
+    bucketVector_->resize(numRows);
+    partition_->extractColumn(
+        bucketColumn_.value(), partitionOffset_, numRows, 0, bucketVector_);
+
+    auto* resultFlatVector = result->asFlatVector<int64_t>();
+    auto* rawValues = resultFlatVector->mutableRawValues();
+    for (auto i = 0; i < numRows; i++) {
+      vector_size_t rowIndex = i + partitionOffset_;
+      if (bucketFlatVector_->isNullAt(i)) {
+        resultFlatVector->setNull(resultOffset + i, true);
+      } else {
+        auto numBuckets = bucketFlatVector_->valueAt(i);
+        VELOX_USER_CHECK(numBuckets >= 1, kBucketErrorString);
+        auto bucketsMoreThanPartition = numBuckets > numPartitionRows_;
+        if (bucketsMoreThanPartition) {
+          rawValues[resultOffset + i] = rowIndex + 1;
+        } else {
+          rawValues[resultOffset + i] =
+              computeBucketValue(rowIndex, computeBucketMetrics(numBuckets));
+        }
+      }
+    }
+  }
+
+  void computeNtileWithConstants(
+      vector_size_t numRows,
+      vector_size_t resultOffset,
+      const VectorPtr& result) {
+    if (numFixedBuckets_.has_value()) {
+      auto* rawValues = result->asFlatVector<int64_t>()->mutableRawValues();
+      if (fixedBucketsMoreThanPartition_) {
+        for (auto i = 0; i < numRows; i++) {
+          rawValues[resultOffset + i] = i + partitionOffset_ + 1;
+        }
+      } else {
+        for (auto i = 0; i < numRows; i++) {
+          rawValues[resultOffset + i] =
+              computeBucketValue(i + partitionOffset_, fixedBucketMetrics_);
```

Comment:
Consider calling computeBucketValue for the whole row vector. Then move the loop inside the if (rowNumber < bucketMetrics.extraBucketsBoundary) branches. I.e. take the "if" out if the loop.